### PR TITLE
Issue 22 base tests

### PR DESCRIPTION
**Pull request by Vitomir2**
* add more ValidatorSet tests and create new System tests;
* make some fixes in the contracts - change names, add imports and other minor bugs;
* rename VestingLib -> VestingPositionLib and RewardPoolContract -> RewardPool;
* create new methods - getValidator, getEpochByBlock and getSortedValidators;
* create new validator flag to check whether it is registered already;
* extract the staking contract related tests in a separate Staking.test.ts file;

### CODE REVIEW

File: test/Hydra/ValidatorSet/ValidatorSet.test.ts [hunk 1]
```diff
@@ -47,57 +56,1747 @@ describe("ValidatorSet", function () {
...
+      const currentEpochId = await systemValidatorSet.currentEpochId();
+
+      expect(storedEpoch.startBlock, "startBlock").to.equal(this.epoch.startBlock);
+      expect(storedEpoch.endBlock, "endBlock").to.equal(this.epoch.endBlock);
+      expect(storedEpoch.epochRoot, "epochRoot").to.equal(hre.ethers.utils.hexlify(this.epoch.epochRoot));
+      expect(currentEpochId, "currentEpochId").to.equal(2);
+    });
+
+    it("should get sorted validators - admin", async function () {
+      const { validatorSet } = await loadFixture(this.fixtures.commitEpochTxFixture);
```
Comment (R-Santev) [inline test/Hydra/ValidatorSet/ValidatorSet.test.ts:297]: Sorted validators can be replaced with simpler function that only get all active validators without sort

File: test/Hydra/mochaContext.ts [hunk 2]
```diff
@@ -1,40 +1,121 @@
...
+      pubkey: [BigNumberish, BigNumberish, BigNumberish, BigNumberish];
+      signature: [BigNumberish, BigNumberish];
+      stake: BigNumberish;
+    };
   }
 }
 
-export function initValidators(accounts: SignerWithAddress[], num: number = 4) {
-  if (num > accounts.length) {
+export function initValidators(accounts: SignerWithAddress[], from: number = 0, to: number = 4) {
```
Comment (R-Santev) [inline test/Hydra/mochaContext.ts:110]: why is this function declared in the mochaContext file?

File: test/Hydra/ValidatorSet/Staking.test.ts [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+    it("should place in withdrawal queue", async function () {
+      const { validatorSet } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
+      await validatorSet.connect(this.signers.validators[0]).unstake(this.minStake.mul(2));
+
+      expect(await validatorSet.pendingWithdrawals(this.signers.validators[0].address)).to.equal(this.minStake.mul(2));
+      expect(await validatorSet.withdrawable(this.signers.validators[0].address)).to.equal(0);
+    });
+  });
+}
```
Comment (R-Santev) [inline test/Hydra/ValidatorSet/Staking.test.ts:211]: Tests for vested staking are not implemented

File: contracts/Hydra/ValidatorSet/modules/Staking/IStaking.sol [hunk 4]
```diff
@@ -37,4 +37,12 @@ interface IStaking {
...
      */
     function setCommission(uint256 newCommission) external;
+
+    /**
+     * @notice Gets first n active validators sorted by total stake.
+     * @param n Desired number of validators to return
+     * @return Returns array of addresses of first n active validators sorted by total stake,
+     * or fewer if there are not enough active validators
+     */
+    function sortedValidators(uint256 n) external view returns (address[] memory);
```
Comment (R-Santev) [inline contracts/Hydra/ValidatorSet/modules/Staking/IStaking.sol:47]: Replace with just active validators. Ping me for more information.

**Review by R-Santev — CHANGES_REQUESTED** [review]
Overal looks amazing. There are just a few things to be improved and added.

File: test/Hydra/ValidatorSet/ValidatorSet.test.ts [hunk 5]
```diff
@@ -47,57 +56,1747 @@ describe("ValidatorSet", function () {
...
+        validator.blsKey.map((x) => x.toHexString()),
+        "blsKey"
+      ).to.deep.equal(this.validatorInit.pubkey);
+      expect(await systemValidatorSet.balanceOf(adminAddress), "balanceOf").to.equal(this.minStake.mul(2));
+      expect(await systemValidatorSet.totalDelegationOf(adminAddress), "totalDelegationOf").to.equal(
+        this.minStake.mul(2)
+      );
+      expect(validator.commission, "commission").to.equal(0);
+      expect(await systemValidatorSet.bls(), "bls").to.equal(bls.address);
+      // TODO: vito.do: read about erc20vote and how it works. Its delegate works differently from our delegate. Then, I will know why totalSupplyAt returns 0
```
[the remark this answers is not in the document]
Comment (Vitomir2) [reply test/Hydra/ValidatorSet/ValidatorSet.test.ts:183]: What do you mean by "fix" test ?

File: test/Hydra/mochaContext.ts (hunk 2 above)
Comment (Vitomir2) [reply test/Hydra/mochaContext.ts:110]: I had to move it into the helper. Will do.

File: test/Hydra/ValidatorSet/ValidatorSet.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vitomir2) [reply test/Hydra/ValidatorSet/ValidatorSet.test.ts]: done

**Comment by Vitomir2** [comment]
Fixed everything related to the comments and even added some stuff to the contracts.

* add the getValidators method to retrieve all the available validators;
* create a new flag for the validator to check when it is registered or not and throw custom error when the validator is already registered;
* add tests for the new method and error that it could throw;
* create new helper methods - registerValidator, findProperRPSIndex, and update the helper that commits multiple epochs;
* adapt the vested staking tests;
* re-arrange the arguments to match the parameters when calling  _calculateValidatorAndDelegatorShares;
* fix typo veting -> vesting;
* delete unused commented code;

File: test/Hydra/ValidatorSet/Staking.test.ts [hunk 6]
```diff
@@ -208,4 +207,277 @@ export function RunStakingTests(): void {
...
+    describe("openVestedPosition()", function () {
+      it("should open vested position", async function () {
+        const { validatorSet, systemValidatorSet, rewardPool } = await loadFixture(
+          this.fixtures.stakedValidatorsStateFixture
+        );
+
+        await registerValidator(validatorSet, this.signers.governance, staker);
+        const stakerValidatorSet = validatorSet.connect(staker);
+        const tx = await stakerValidatorSet.openVestedPosition(VESTING_DURATION_WEEKS, {
+          value: this.minDelegation,
```
Comment (R-Santev) [inline test/Hydra/ValidatorSet/Staking.test.ts:228]: minDelegation is used for stake action. Maybe better to create a minStake variable?

File: test/Hydra/ValidatorSet/Staking.test.ts [hunk 7]
```diff
@@ -208,4 +207,277 @@ export function RunStakingTests(): void {
...
+
+    describe("Top-up staking position with stake()", function () {
+      it("should top-up staking position", async function () {
+        const { stakerValidatorSet, systemValidatorSet, rewardPool } = await loadFixture(
+          this.fixtures.newVestingValidatorFixture
+        );
+
+        const tx = await stakerValidatorSet.stake({ value: this.minDelegation });
+        const vestingData = await rewardPool.positions(staker.address);
+        if (!tx.blockNumber) {
```
Comment (R-Santev) [inline test/Hydra/ValidatorSet/Staking.test.ts:275]: There is no need of blockNumber checks in my opinion.

File: test/Hydra/ValidatorSet/ValidatorSet.test.ts [hunk 8]
```diff
@@ -293,12 +291,10 @@ describe("ValidatorSet", function () {
       expect(currentEpochId, "currentEpochId").to.equal(2);
     });
 
-    it("should get sorted validators - admin", async function () {
+    it("should all active validators - admin", async function () {
```
Comment (R-Santev) [inline test/Hydra/ValidatorSet/ValidatorSet.test.ts:294]: "get" is missed?

**Review by R-Santev — CHANGES_REQUESTED** [review]

**Comment by Vitomir2** [comment]
Resolved all the remarks.

**Review by R-Santev — APPROVED** [review]

**Pull request merged**